csrc/python_frontend/python_bindings.cpp

```diff
@@ -1153,6 +1153,45 @@ void initNvFuserPythonBindings(PyObject* module) {
   NVFUSER_PYTHON_BINDING_BINARY_OP_TENSORS_ONLY("matmul", matmul)
 #undef NVFUSER_PYTHON_BINDING_BINARY_OP_TENSORS_ONLY
 
+  nvf_ops.def(
+      "linear",
+      [](FusionDefinition::Operators& self,
+         Tensor arg1,
+         Tensor arg2,
+         std::optional<Tensor> bias = std::nullopt) -> Tensor {
+        FUSER_PERF_SCOPE("Operators.linear");
+        NVF_CHECK(
+            self.validUse(), "Attempting to add to a completed definition!");
+        FusionDefinition* fd = self.fusion_definition;
+        Tensor output = fd->defineTensor(arg1.dims);
+
+        if (bias.has_value()) {
+          fd->defineRecord(
+              new OpRecord<TensorView*, TensorView*, TensorView*, TensorView*>(
+                  {fd->recordingState(arg1()),
+                   fd->recordingState(arg2()),
+                   fd->recordingState(bias.value()())},
```

Comment:
Is it not possible to have a missing value for bias here? i.e. can we just insert `nullptr` here instead of having two separate `OpRecord` definitions?

Reply:
Nevermind. Of course these are `State`s, so there's no way to just say this argument should be nullptr.

Reply:
Not quite a nullptr, but something like this:

```
 auto bias_state = bias.has_value()
            ? fd->recordingState(bias.value()())
            : State(0, serde::StateType::None);
```

And then we handle the None state. This pattern exists in our current `python_bindings.cpp`, and to make it work we'll need to create a new class like `OpRecord` for example, `LinearOpRecord`. Creating another class was seen as not really desirable. (https://github.com/NVIDIA/Fuser/blob/5c674cde66101a497bbaeed097624a60eefae07a/csrc/python_frontend/python_bindings.cpp#L2699-L2704)


The limitations we hit are:

```
struct OpRecord : RecordFunctor {
  OpRecord(
      std::vector<State> _args,
      std::vector<State> _outputs,
      std::string _name,
      serde::RecordType record_type,
      std::function<OutType(ArgTypes...)> fusion_op)
      : RecordFunctor(
            std::move(_args),
            std::move(_outputs),
            _name,
            record_type),
        fusion_op_(fusion_op)
```
Where the fusion_op has the same number input args as the template args. We thought about getting around this with `std::optional` but this doesn't quite work out because of:
https://github.com/NVIDIA/Fuser/blob/5c674cde66101a497bbaeed097624a60eefae07a/csrc/python_frontend/fusion_record.h#L302-L307

Another point to note is that, the two OpRecords will have different `serde::RecordType`, Binary and Ternary.

Reply:
Actually there is `StateType::None`:
https://github.com/NVIDIA/Fuser/blob/9c220186f35a8478dbcd95f178f3749ebfaa1369/csrc/serde/fusion_cache.fbs#L16-L22
What happens if you just create one of these objects for the null bias if `!bias.hasValue()`? (don't merge this change)
```suggestion
                   fd->recordingState(bias.has_value() ? bias.value()() : State{.index=0, .stype=serde::StateType::None())},
```

Reply:
(you beat me to it). I don't understand why it requires a new class. `StateType::None()` should just imply we pass a `nullptr` for that argument right? So the number of args should match.

Reply:
It seems like this would be fixed if `fd.getFusionState` would return `nullptr` whenever `stype=None`.

Reply:
@jjsjann123, @rdspring1 , @kevinstephano Tagging you guys since you may worked on this.
What would be the best way to pass in optional arguments to an Op like Linear.
My ideas were:
1. Create a new Record, like LinearOpRecord (cons: more work and code bloat)
2. Different `linear` functions (possibly via a lambda as @jacobhinkle pointed).
3. Is there another way?

I don't want to change the general way we handle default/optional arguments in our python fronted in this PR though.

Reply:
> It seems like this would be fixed if fd.getFusionState would return nullptr whenever stype=None.
I am going to try this out.

Reply:
I know it is slightly ugly to have two `linear` functions on the C++ side but I would like to move forward with just using a 2 argument and a 3 argument signature so we can move forward.

Reply:
Not against @kevinstephano 's suggestion on keeping it as-is for faster movement. @jacobhinkle 's suggestion doesn't seem to suggest lot more work... We should at least follow up with that as a clean up.

Reply:
Can we have optional arguments in our cpp Fusion Definition? 

I was seeing `RuntimeError: kernel->inputs().size() <= args.size()` with this PRs ToT.
Here was the failing case:
```python
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1], contiguity=[True, True], dtype=DataType.Half, is_cpu=False, stride_order=[1, 0])
    T1 = fd.define_tensor(shape=[-1, -1], contiguity=[True, True], dtype=DataType.Half, is_cpu=False, stride_order=[1, 0])
    T2 = fd.define_tensor(shape=[-1], contiguity=[True], dtype=DataType.Half, is_cpu=False, stride_order=[0])
    T3 = fd.ops.linear(T0, T1, T2)
    fd.add_output(T3)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randn((192,), dtype=torch.float16, device='cuda:0').as_strided((24, 8), (8, 1)),
    torch.randn((128,), dtype=torch.float16, device='cuda:0').as_strided((16, 8), (8, 1)),
]
fd.execute(inputs)
```

Reply:
You are missing bias in `inputs` :wink:

Reply:
you defined three input tensor in your FusionDefinition.

Reply:
That error looks legit. If we provide a bias in the fusion definition, we need to also supply it to `fd.execute`.

Reply:
That errors is fixed, let me update the PR. This showed since I merged some of Jacob's comments on the browser

Reply:
@rdspring1 can you try it again. It should be fixed now.